File: apps/desktop/src/renderer/regions.test.ts
import assert from "node:assert/strict";
import test from "node:test";
import {
  buildServerRetryOrder,
  groupRegions,
  loadOf,
  orderByRecent,
  pickNode,
  promoteRecent,
  regionKeyOf,
  regionLoad,
  regionOfServer
} from "./regions.ts";

const server = (id: string, name: string, load: number | null = 0, country = ""): ServerInfo => ({
  id,
  name,
  region: name,
  country,
  load,
  cloak: { remoteHost: `${id}.example`, uid: "uid", publicKey: "pk" }
});

test("regionKeyOf strips a trailing numeric suffix", () => {
  assert.equal(regionKeyOf(server("eu-central-1", "Amsterdam")), "eu-central");
  assert.equal(regionKeyOf(server("eu-central-12", "Amsterdam")), "eu-central");
});

test("regionKeyOf leaves ids without a numeric suffix alone", () => {
  assert.equal(regionKeyOf(server("amsterdam", "Amsterdam")), "amsterdam");
  assert.equal(regionKeyOf(server("eu-west-a", "London")), "eu-west-a");
});

test("groupRegions folds sibling nodes into one region, in hub order", () => {
  const regions = groupRegions([
    server("eu-central-1", "Amsterdam"),
    server("eu-west-1", "London"),
    server("eu-central-2", "Amsterdam")
  ]);

  assert.deepEqual(regions.map((r) => r.key), ["eu-central", "eu-west"]);
  assert.deepEqual(regions[0].nodes.map((n) => n.id), ["eu-central-1", "eu-central-2"]);
  assert.equal(regions[0].name, "Amsterdam");
  assert.equal(regions[1].nodes.length, 1);
});

test("groupRegions falls back to the key when a node has no name", () => {
  const [region] = groupRegions([server("eu-central-1", "")]);
  assert.equal(region.name, "eu-central");
});

test("pickNode returns the lowest-load node", () => {
  const [region] = groupRegions([
    server("eu-central-1", "Amsterdam", 61),
    server("eu-central-2", "Amsterdam", 22),
    server("eu-central-3", "Amsterdam", 47)
  ]);
  assert.equal(pickNode(region).id, "eu-central-2");
  assert.equal(regionLoad(region), 22);
});

test("a node reporting load beats one reporting none", () => {
  const [region] = groupRegions([
    server("eu-central-1", "Amsterdam", null),
    server("eu-central-2", "Amsterdam", 90)
  ]);
  assert.equal(pickNode(region).id, "eu-central-2");
});

test("regionLoad is null when no node in the region reports one", () => {
  const [region] = groupRegions([server("eu-west-1", "London", null)]);
  assert.equal(regionLoad(region), null);
  assert.equal(loadOf(region.nodes[0]), 100);
});

test("regionOfServer finds the region holding a node id", () => {
  const regions = groupRegions([
    server("eu-central-1", "Amsterdam"),
    server("us-east-1", "New York")
  ]);
  assert.equal(regionOfServer(regions, "us-east-1")?.key, "us-east");
  assert.equal(regionOfServer(regions, "nope-1"), undefined);
});

test("orderByRecent puts recents first and keeps the rest in hub order", () => {
  const regions = groupRegions([
    server("eu-central-1", "Amsterdam"),
    server("eu-west-1", "London"),
    server("us-east-1", "New York")
  ]);
  const ordered = orderByRecent(regions, ["us-east", "eu-west"]);
  assert.deepEqual(ordered.map((r) => r.key), ["us-east", "eu-west", "eu-central"]);
});

test("orderByRecent ignores recents that no longer exist", () => {
  const regions = groupRegions([server("eu-west-1", "London")]);
  assert.deepEqual(orderByRecent(regions, ["gone", "eu-west"]).map((r) => r.key), ["eu-west"]);
});

test("promoteRecent moves a key to the front without duplicating it", () => {
  assert.deepEqual(promoteRecent(["a", "b", "c"], "c"), ["c", "a", "b"]);
  assert.deepEqual(promoteRecent(["a", "b"], "z"), ["z", "a", "b"]);
});

test("promoteRecent caps the list", () => {
  assert.deepEqual(promoteRecent(["a", "b", "c"], "z", 2), ["z", "a"]);
});

test("buildServerRetryOrder exhausts the selected region before later regions", () => {
  const servers = [
    server("eu-central-1", "Amsterdam", 70),
    server("us-east-1", "New York", 5),
    server("eu-central-2", "Amsterdam", 20),
    server("eu-central-3", "Amsterdam", null),
    server("us-east-2", "New York", 10),
    server("ap-south-1", "Mumbai", 1)
  ];

  assert.deepEqual(buildServerRetryOrder(servers, "eu-central-1"), [
    "eu-central-1",
    "eu-central-2",
    "eu-central-3",
    "us-east-1",
    "us-east-2",
    "ap-south-1"
  ]);
  assert.deepEqual(servers.map((item) => item.id), [
    "eu-central-1",
    "us-east-1",
    "eu-central-2",
    "eu-central-3",
    "us-east-2",
    "ap-south-1"
  ]);
});

test("buildServerRetryOrder keeps the requested server when it is not in the snapshot", () => {
  assert.deepEqual(buildServerRetryOrder([server("eu-west-1", "London")], "gone-1"), ["gone-1"]);
});

test("buildServerRetryOrder continues with the next region and wraps around", () => {
  const servers = [
    server("eu-west-1", "London", 20),
    server("us-east-1", "New York", 30),
    server("ap-south-1", "Mumbai", 40)
  ];

  assert.deepEqual(buildServerRetryOrder(servers, "us-east-1"), [
    "us-east-1",
    "ap-south-1",
    "eu-west-1"
  ]);
});

File: apps/android/core/src/test/kotlin/org/pangeavpn/core/RegionsTest.kt
package org.pangeavpn.core

import kotlin.random.Random
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNull
import kotlin.test.assertTrue
import org.pangeavpn.core.model.Server

private fun server(id: String, load: Int? = null, name: String = "Node", country: String = "GB") =
    Server(id = id, name = name, region = "eu", country = country, load = load)

class RegionsTest {

    @Test
    fun `region key strips the numeric node suffix`() {
        assertEquals("eu-central", regionKeyOf(server("eu-central-1")))
        assertEquals("eu-central", regionKeyOf(server("eu-central-12")))
        assertEquals("standalone", regionKeyOf(server("standalone")))
    }

    @Test
    fun `grouping preserves first seen order and collects nodes`() {
        val regions = groupRegions(
            listOf(server("eu-1"), server("us-1"), server("eu-2")),
        )
        assertEquals(listOf("eu", "us"), regions.map { it.key })
        assertEquals(2, regions.first().nodes.size)
    }

    @Test
    fun `a node that reports load beats one that does not`() {
        val region = groupRegions(listOf(server("eu-1", load = null), server("eu-2", load = 90))).first()
        assertEquals("eu-2", pickNode(region).id)
    }

    @Test
    fun `region picks its lowest load node`() {
        val region = groupRegions(listOf(server("eu-1", 80), server("eu-2", 10), server("eu-3", 40))).first()
        assertEquals("eu-2", pickNode(region).id)
        assertEquals(10, regionLoad(region))
    }

    @Test
    fun `region load is null when no node reports`() {
        val region = groupRegions(listOf(server("eu-1"), server("eu-2"))).first()
        assertNull(regionLoad(region))
    }

    @Test
    fun `random pick prefers lightly loaded servers`() {
        val servers = listOf(server("a", 10), server("b", 90), server("c", 20))
        repeat(50) { seed ->
            val picked = pickRandomServer(servers, Random(seed))
            assertTrue(picked?.id in setOf("a", "c"), "congested node b should not be picked")
        }
    }

    @Test
    fun `random pick falls back when everything is congested`() {
        val servers = listOf(server("a", 90), server("b", 95))
        assertTrue(pickRandomServer(servers, Random(1))?.id in setOf("a", "b"))
    }

    @Test
    fun `random pick falls back when nothing reports load`() {
        val servers = listOf(server("a"), server("b"))
        assertTrue(pickRandomServer(servers, Random(1))?.id in setOf("a", "b"))
    }

    @Test
    fun `random pick of an empty list is null`() {
        assertNull(pickRandomServer(emptyList()))
    }

    @Test
    fun `selection keeps the in-session pick while it is listed`() {
        val visible = listOf(server("a"), server("b"))
        assertEquals("b", resolveSelection(visible, "b", "a"))
    }

    @Test
    fun `selection falls back to the last connected server`() {
        val visible = listOf(server("a"), server("b"))
        assertEquals("a", resolveSelection(visible, "gone", "a"))
    }

    @Test
    fun `selection empties when neither is listed so connect rolls a random one`() {
        val visible = listOf(server("a"))
        assertEquals("", resolveSelection(visible, "gone", "also-gone"))
    }

    @Test
    fun `region of server finds the owning region`() {
        val regions = groupRegions(listOf(server("eu-1"), server("us-1")))
        assertEquals("us", regionOfServer(regions, "us-1")?.key)
        assertNull(regionOfServer(regions, "nope"))
    }
}
